Classes/FairPlayDRM/FPSAssetLoaderDelegate.swift

```diff
@@ -0,0 +1,174 @@
+        let delegate = FPSAssetLoaderDelegate.init(storage: storage)
+        
+        asset.resourceLoader.setDelegate(delegate, queue: resourceLoadingRequestQueue)
+        asset.resourceLoader.preloadsEligibleContentKeys = true
+        
+        return delegate
+    }
+    
+    func prepareAndSendContentKeyRequest(resourceLoadingRequest: AVAssetResourceLoadingRequest) {
+        
+        guard let assetId = resourceLoadingRequest.request.url?.host else {
+            PKLog.error("No asset id")
+            return
+        }
+        
+        // Check if this reuqest is the result of a potential AVAssetDownloadTask.
+        if #available(iOS 10.0, *), shouldPersist {
+            if let cir = resourceLoadingRequest.contentInformationRequest {
+                cir.contentType = AVStreamingKeyDeliveryPersistentContentKeyType
+            }
+        }
+        
+        guard let helper = FPSLicenseHelper.init(assetId: assetId, params: self.drmData, 
+                                                 dataStore: self.storage, forceDownload: self.forceDownload) else { return }
+        
+        helper.handleLicenseRequest(FPSResourceLoadingKeyRequest(resourceLoadingRequest)) { (error) in
+            self.done?(error)
+        }
+    }
+
+    func shouldLoadOrRenewRequestedResource(resourceLoadingRequest: AVAssetResourceLoadingRequest) -> Bool {
+        
+        guard let url = resourceLoadingRequest.request.url else {
+            return false
+        }
+        
+        // AssetLoaderDelegate only should handle FPS Content Key requests.
+        if url.scheme != FPSAssetLoaderDelegate.customScheme {
+            return false
+        }
+        
+        FPSAssetLoaderDelegate.resourceLoadingRequestQueue.async {
+            self.prepareAndSendContentKeyRequest(resourceLoadingRequest: resourceLoadingRequest)
+        }
+        
+        return true
+    }
+}
+
+//MARK:- AVAssetResourceLoaderDelegate protocol methods extension
+extension FPSAssetLoaderDelegate: AVAssetResourceLoaderDelegate {
+    
+    func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForLoadingOfRequestedResource loadingRequest: AVAssetResourceLoadingRequest) -> Bool {
+        
+        PKLog.trace("\(#function) was called in AssetLoaderDelegate with loadingRequest: \(loadingRequest)")
+        
+        return shouldLoadOrRenewRequestedResource(resourceLoadingRequest: loadingRequest)
+    }
+    
+    func resourceLoader(_ resourceLoader: AVAssetResourceLoader, shouldWaitForRenewalOfRequestedResource renewalRequest: AVAssetResourceRenewalRequest) -> Bool {
+        
+        PKLog.trace("\(#function) was called in AssetLoaderDelegate with renewalRequest: \(renewalRequest)")
+        
+        return shouldLoadOrRenewRequestedResource(resourceLoadingRequest: renewalRequest)
+    }
+}
+
+class FPSResourceLoadingKeyRequest: FPSLicenseRequest {
```

Comment:
If this class will stay in this file, can you please add "//MARK: -" to separate it.